### PR TITLE
Add eventlistener describe command

### PR DESCRIPTION
**Pull request by savitaashture**
Part of #594

Signed-off-by: Savita Ashture sashture@redhat.com
# Changes
Add Describe Command for EventListener

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Run the code checkers with `make check`
- [x] Regenerate the manpages, docs and go formatting with `make generated`
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/cli/blob/master/CONTRIBUTING.md)
for more details._

# Release Notes

```
Add tkn el desc command to describe a EventListener
```

### CODE REVIEW

**Comment by piyush-garg** [comment]
@savitaashture Why we are showing some fields in the describe command output as json

File: pkg/cmd/eventlistener/describe.go [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+	if err = tparsed.Execute(w, data); err != nil {
+		fmt.Fprintf(s.Err, "Failed to execute template \n")
+		return err
+	}
+	return w.Flush()
+}
+
+func getEventListenerTrigger(listener v1alpha1.EventListener) string {
+	var out bytes.Buffer
+	resourceTemplate, err := json.Marshal(listener.Spec.Triggers)
```
Comment (danielhelfand) [inline pkg/cmd/eventlistener/describe.go:176]: Not sure if there is a reason for the JSON format versus YAML here. Maybe we could support both formats? I would personally prefer YAML for this.

File: pkg/cmd/eventlistener/describe.go [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+	w := tabwriter.NewWriter(s.Out, 0, 5, 3, ' ', tabwriter.TabIndent)
+	tparsed := template.Must(template.New("Describe EventListener").Funcs(funcMap).Parse(describeTemplate))
+	if err = tparsed.Execute(w, data); err != nil {
+		fmt.Fprintf(s.Err, "Failed to execute template \n")
+		return err
+	}
+	return w.Flush()
+}
+
+func getEventListenerTrigger(listener v1alpha1.EventListener) string {
```
Comment (danielhelfand) [inline pkg/cmd/eventlistener/describe.go:174]: Maybe return `(string, error)` here?

File: pkg/cmd/eventlistener/describe.go [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+	var out bytes.Buffer
+	resourceTemplate, err := json.Marshal(listener.Spec.Triggers)
+	if err != nil {
+		log.Printf("json marshal failed with error: %v", err)
+		return ""
```
Comment (danielhelfand) [inline pkg/cmd/eventlistener/describe.go:179]: Returning an error here would be preferred/would make this easier to test.

File: pkg/cmd/eventlistener/describe.go [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+	return printer.PrintObject(w, el, f)
+}
+
+func printEventListenerDescription(s *cli.Stream, p cli.Params, ttname string) error {
+	cs, err := p.Clients()
+	if err != nil {
+		return fmt.Errorf("failed to create tekton client")
+	}
+
+	el, err := cs.Triggers.TriggersV1alpha1().EventListeners(p.Namespace()).Get(ttname, metav1.GetOptions{})
```
Comment (danielhelfand) [inline pkg/cmd/eventlistener/describe.go:146]: `elName` instead of `ttname`

File: pkg/cmd/eventlistener/describe.go [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+{{- $value := getEventListenerName .EventListener }}
+{{-  if ne $value "" }}
+{{decorate "green" "EventListnerServiceName"}}:{{ $value }}
+{{- end }}
+
+`
+
+func describeCommand(p cli.Params) *cobra.Command {
+	f := cliopts.NewPrintFlags("describe")
+	eg := `Describe a EventListener of name 'foo' in namespace 'bar':
```
Comment (danielhelfand) [inline pkg/cmd/eventlistener/describe.go:70]:
```suggestion
	eg := `Describe an EventListener of name 'foo' in namespace 'bar':
```

File: pkg/cmd/eventlistener/describe.go [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+
+or
+
+   tkn el desc foo -n bar
+`
+
+	c := &cobra.Command{
+		Use:     "describe",
+		Aliases: []string{"desc"},
+		Short:   "Describes a eventlistener in a namespace",
```
Comment (danielhelfand) [inline pkg/cmd/eventlistener/describe.go:82]:
```suggestion
		Short:   "Describe EventListeners in a namespace",
```

File: pkg/cmd/eventlistener/describe.go [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+	el.GetObjectKind().SetGroupVersionKind(
+		schema.GroupVersionKind{
+			Version: "triggers.tekton.dev",
+			Kind:    "EventListener",
+		})
+
+	return printer.PrintObject(w, el, f)
+}
+
+func printEventListenerDescription(s *cli.Stream, p cli.Params, ttname string) error {
```
Comment (danielhelfand) [inline pkg/cmd/eventlistener/describe.go:140]: `elName` instead of `ttname`

File: pkg/cmd/eventlistener/describe_test.go [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+
+	eventListener := Command(p)
+	out, err := test.ExecuteCommand(eventListener, "desc", "bar", "-n", "ns")
+	if err == nil {
+		t.Errorf("Error expected here")
+	}
+	golden.Assert(t, out, fmt.Sprintf("%s.golden", t.Name()))
+}
+
+func TestEventListenerDescribe_Empty(t *testing.T) {
```
Comment (danielhelfand) [inline pkg/cmd/eventlistener/describe_test.go:58]:
```suggestion
func TestEventListenerDescribe_NoArgProvided(t *testing.T) {
```

File: pkg/cmd/eventlistener/describe_test.go [hunk 9]
```diff
@@ -0,0 +1,205 @@
+// Copyright © 2020 The Tekton Authors.
```
Comment (danielhelfand) [inline pkg/cmd/eventlistener/describe_test.go:1]: There should be a test added for an `EventListener` with only a `Name`/`Namepsace`. Basically the minimum required fields specified.

File: pkg/cmd/eventlistener/describe.go [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+{{- $value := getEventListenerTrigger .EventListener }}
+{{-  if ne $value "" }}
+
+{{decorate "resources" ""}}{{decorate "underline bold" "EventListenerTrigger\n"}}
+{{ $value }}
+{{- end }}
+
+{{- $value := getURL .EventListener }}
+{{-  if ne $value "" }}
+{{decorate "green" "URL"}}:{{ $value }}
```
Comment (danielhelfand) [inline pkg/cmd/eventlistener/describe.go:58]: There should be a space here:

```suggestion
{{decorate "green" "URL"}}: {{ $value }}
```

File: pkg/cmd/eventlistener/describe.go [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+{{- end }}
+
+{{- $value := getEventListenerName .EventListener }}
+{{-  if ne $value "" }}
+{{decorate "green" "EventListnerServiceName"}}:{{ $value }}
```
Comment (danielhelfand) [inline pkg/cmd/eventlistener/describe.go:63]: There should be a space here:

```suggestion
{{decorate "green" "EventListnerServiceName"}}: {{ $value }}
```

File: pkg/cmd/eventlistener/describe.go [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+{{- end }}
+
+{{- if ne .EventListener.Spec.ServiceType "" }}
+{{decorate "bold" "Service Type"}}:	{{ .EventListener.Spec.ServiceType }}
+{{- end }}
+
+{{- $value := getEventListenerTrigger .EventListener }}
+{{-  if ne $value "" }}
+
+{{decorate "resources" ""}}{{decorate "underline bold" "EventListenerTrigger\n"}}
```
Comment (danielhelfand) [inline pkg/cmd/eventlistener/describe.go:52]:
```suggestion
{{decorate "resources" ""}}{{decorate "underline bold" "EventListenerTrigger(s): \n"}}
```

File: pkg/cmd/eventlistener/describe.go (hunk 3 above)
Comment (savitaashture) [reply pkg/cmd/eventlistener/describe.go]: Handled Please take a look

File: pkg/cmd/eventlistener/describe.go (hunk 2 above)
Comment (savitaashture) [reply pkg/cmd/eventlistener/describe.go]: I think template `Funcs` doesn't allow to return more than one type
```
can't install method/function "xxx" with 2 results
```

File: pkg/cmd/eventlistener/describe.go (hunk 1 above)
Comment (savitaashture) [reply pkg/cmd/eventlistener/describe.go]: Modified 
Thank you

File: pkg/cmd/eventlistener/describe_test.go (hunk 9 above)
Comment (savitaashture) [reply pkg/cmd/eventlistener/describe_test.go:1]: Added Testcase

File: pkg/cmd/eventlistener/describe.go [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+{{-  if ne $value "" }}
+{{decorate "bold" "EventListnerServiceName"}}:	{{ $value }}
+{{- end }}
+
+{{- if eq (len .EventListener.Spec.Triggers) 0 }}
+
+ No EventListenerTriggers
+{{- else }}
+
+{{decorate "resources" ""}}{{decorate "underline bold" "Event Listener Triggers\n"}}
```
Comment (danielhelfand) [inline pkg/cmd/eventlistener/describe.go:62]: Why is this needed?

File: pkg/cmd/eventlistener/describe.go [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+{{- range $v := .EventListener.Spec.Triggers }}
+{{decorate "cyan" "EventListenerTrigger\n"}}
```
Comment (danielhelfand) [inline pkg/cmd/eventlistener/describe.go:64]: Make this plural:

```suggestion
{{decorate "cyan" "EventListenerTriggers\n"}}
```

File: pkg/cmd/eventlistener/describe.go [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+}
+
+func getBinding(binding []*v1alpha1.EventListenerBinding) []*v1alpha1.EventListenerBinding {
+	return binding
+}
+
+func getInterceptors(interceptors []*v1alpha1.EventInterceptor) string {
+	resourceTemplate, err := yaml.Marshal(interceptors)
+	if err != nil {
+		return "json marshal failed with error: " + err.Error()
```
Comment (danielhelfand) [inline pkg/cmd/eventlistener/describe.go:204]:
```suggestion
		return "yaml marshal failed with error: " + err.Error()
```

File: pkg/cmd/eventlistener/testdata/TestEventListenerDescribe_WithCELInterceptors.golden [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+ tb3             TriggerBinding   v1alpha1
+ TEMPLATE NAME   APIVERSION
+ tt1             v1alpha1
+ INTERCEPTORS
+- cel:
+    filter: body.value == 'test'
+    overlays:
+    - expression: '''testing'''
+      key: value
+
```
Comment (danielhelfand) [inline pkg/cmd/eventlistener/testdata/TestEventListenerDescribe_WithCELInterceptors.golden:22]: Can we find a way to avoid the extra line break with interceptors?

File: pkg/cmd/eventlistener/describe.go [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+
+or
+
+   tkn el desc foo -n bar
+`
+
+	c := &cobra.Command{
+		Use:     "describe",
+		Aliases: []string{"desc"},
+		Short:   "Describes eventlistener in a namespace",
```
Comment (danielhelfand) [inline pkg/cmd/eventlistener/describe.go:104]:
```suggestion
		Short:   "Describe EvenListeners in a namespace",
```

File: pkg/cmd/eventlistener/describe.go (hunk 13 above)
Comment (savitaashture) [reply pkg/cmd/eventlistener/describe.go:62]: If there are multiple triggers in order to differentiate those added that perticular line https://github.com/tektoncd/cli/pull/933/files?file-filters%5B%5D=.golden#diff-9f7e468100cafaa9eb9cb1a9a0ab0088R4-R46 

But i think i can remove that as you suggested

File: pkg/cmd/eventlistener/describe.go [hunk 18]
```diff
@@ -0,0 +1,219 @@
...
+{{decorate "bold" "EventListnerServiceName"}}:	{{ $value }}
+{{- end }}
+
+{{- if eq (len .EventListener.Spec.Triggers) 0 }}
+
+ No EventListenerTriggers
+{{- else }}
+{{ " " }}
+{{- range $v := .EventListener.Spec.Triggers }}
+{{decorate "cyan" "EventListenerTriggers\n"}}
```
Comment (danielhelfand) [inline pkg/cmd/eventlistener/describe.go:63]: Sorry, I missed this, but we shouldn't use color here. Probably just make it bold.

File: pkg/cmd/eventlistener/testdata/TestEventListenerDescribe_WithCELInterceptors.golden (hunk 16 above)
Comment (savitaashture) [reply pkg/cmd/eventlistener/testdata/TestEventListenerDescribe_WithCELInterceptors.golden:20]: I have removed extra line

File: pkg/cmd/eventlistener/describe.go (hunk 18 above)
Comment (savitaashture) [reply pkg/cmd/eventlistener/describe.go]: Done

File: pkg/cmd/eventlistener/describe.go [hunk 19]
```diff
@@ -0,0 +1,219 @@
...
+
+or
+
+   tkn el desc foo -n bar
+`
+
+	c := &cobra.Command{
+		Use:     "describe",
+		Aliases: []string{"desc"},
+		Short:   "Describe EvenListeners in a namespace",
```
Comment (danielhelfand) [inline pkg/cmd/eventlistener/describe.go:102]: Everything looks great and thanks! Just one last nit:

```suggestion
		Short:   "Describe EventListeners in a namespace",
```
